repo.go

```diff
@@ -862,26 +863,34 @@ func (r *Repo) AddOrUpdateSignature(roleFilename string, signature data.Signatur
 		return err
 	}
 
-	// Add or update signature.
+	// Check signature on signed meta.
+	// We temporarily make this signature the only signature on the data and modify
+	// the db verification to allow a single signature to suffice.
+	// This way, we test if this new signature is valid.
+	oldSigs := s.Signatures
+	s.Signatures = []data.Signature{
+		signature,
+	}
+	for _, db := range dbs {
+		dbRole := db.GetRole(role)
+		// NOTE: We modify the database used to verify to have a threshold of 1.
+		// This way, we are testing whether this singly added new signature is valid.
```

Comment:
This is a neat hack, but also provides a good example of they the database model is often more trouble than it's worth. I do see the value in providing a high-level API to interact with the evaluated state of the system, but we must also have the ability to operate on and evaluate individual metadata to improve clarity of the code.

This isn't a criticism of the change here, just continuing to collect evidence for a simplification effort.

Reply:
In all honesty, I reflected on what I'd done on Friday and realized I was making it way too complicated for anyone else to read that code.

The better thing would be if I just factored out the verification logic from the DBs (the reason why I couldn't just use a verifier and go is because I'd have to dupe the canonicalization logic). So I will do that!

Reply:
That sounds great, thanks! 
(I've moved the PR to draft for now while you work on that)

Reply:
https://github.com/theupdateframework/go-tuf/pull/418/commits/8af86a0c71cb8630d40f2728398f5853556e1145

This commit implements the change: the only pitfall is that in the DB model, we need to recanonicalize the metadata at each interval.

As a second follow-up, I can consolidate all the places where we `EncodeCanonical` payload (in repo.go we do this to retrieve signed metadata files for offline signing, it sign.go we do this to sign, and in verify.go we do this to verify). This would remove re-canonicalization as part of the sign/verify flow, and put canonicalization logic closer to the `data` API. WDYT?